services/config.go

```
@@ -0,0 +1,264 @@
+		return sleep
+	}
+	go callRandomly(fn, opts.PollInterval, opts.PollJitter, _configService.done, _configService.logger)
+
+	return _configService.Stop, nil
+}
+
+func (cs *configService) updateClientInfo(conf *apipb.ConfigResponse) {
+	cs.clientInfo.ProToken = conf.ProToken
+	cs.clientInfo.Country = conf.Country
+	cs.clientInfo.Ip = conf.Ip
+}
+
+func (cs *configService) Stop() {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	if !cs.running {
+		return
+	}
+
+	close(cs.done)
+	cs.running = false
+}
+
+// fetchConfig fetches the current config from the server and updates the client's config if a change
+// has occurred. It returns the extra sleep time received from the server response and any error that
+// occurred.
+func (cs *configService) fetchConfig() (int64, error) {
+	op := ops.Begin("Fetching config")
+	defer op.End()
+
+	newConf, sleep, err := cs.fetch()
+	if err != nil {
+		return 0, op.FailIf(err)
+	}
+
+	cs.lastFetched = time.Now()
+
+	cs.logger.Debug("Received config")
+	curConf := cs.configHandler.GetConfig()
+	if curConf != nil && !configIsNew(curConf, newConf) {
+		op.Set("config_changed", false)
+		cs.logger.Debug("Config is unchanged")
+		return sleep, nil
+	}
+
+	op.Set("config_changed", true)
+
+	cs.updateClientInfo(newConf)
+	cs.configHandler.SetConfig(newConf)
+
+	return sleep, nil
+}
+
+func (cs *configService) fetch() (*apipb.ConfigResponse, int64, error) {
+	confReq := cs.newRequest()
+	buf, err := proto.Marshal(confReq)
+	if err != nil {
+		return nil, 0, fmt.Errorf("unable to marshal config request: %w", err)
+	}
+
+	resp, sleep, err := post(
+		cs.opts.OriginURL,
+		bytes.NewReader(buf),
+		cs.opts.RoundTripper,
+		cs.opts.UserConfig,
+		cs.logger,
+	)
+	if err != nil {
+		return nil, 0, fmt.Errorf("config request failed: %w", err)
+	}
+	defer resp.Close()
+
+	gzReader, err := gzip.NewReader(resp)
+	if err != nil {
+		return nil, sleep, fmt.Errorf("unable to open gzip reader: %w", err)
+	}
+
+	configBytes, err := io.ReadAll(gzReader)
```

This error is never checked.